doc/design/executor.md

```diff
@@ -0,0 +1,32 @@
+# Executor Desgin Doc
+
+## Overview
+
+`Executor` evaluates a `ProgramDesc`. Essentially, it instantializes Variables and Operators, then run all the operators
+
+```c++
+void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id) {
+  auto& block = pdesc.blocks(block_id);
+  auto& device = device_contexts_[0];
+
+  // Instantiate all the vars in the global scope
+  for (auto& var : block.vars()) {
+    scope->NewVar(var.name());
+  }
+
+  // Run the block
+  Scope& local_scope = scope->NewScope();
+  for (size_t i = 0; i < should_run.size(); ++i) {
```

Comment:
In my mind the executor need to run an OP **as soon as** its dependent OPs are all finished.

The code here is a viable but low performance solution. Should it be in the design doc?

Reply:
sorry, this is a typo.

Reply:
Do we need to put implementation details in the design doc, the current implementation is one way of implementing it :)